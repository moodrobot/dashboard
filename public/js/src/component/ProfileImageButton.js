import React, { PureComponent } from "react";
import PropTypes from "prop-types";

class ProfileImageButton extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div className="vtop">
        <label className="profile-image">
          <input
            type="radio"
            name="default_img"
            value={this.props.image}
            checked={this.props.checked}
            onChange={this.props.handleClick}
          />
          {/*<span className="icon-upload-lg" />*/}
          <img className="icon-upload-sm" src={this.props.image ? this.props.image : "images/upload-icon.png"} />
        </label>
      </div>
    );
  }
}

ProfileImageButton.propTypes = {
  image: PropTypes.string,
  checked: PropTypes.bool,
  handleClick: PropTypes.func.isRequired
};

export default ProfileImageButton;
