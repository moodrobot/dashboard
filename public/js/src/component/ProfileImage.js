import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import axios from "axios";

import ProfileImageButton from "./ProfileImageButton";

class ProfileImage extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      uploadLoader: "off",
      removeLoader: "off",
      message: "",
      chosen_image: ""
    };

    this.chosen_image = "";
    this.handleClick = this.handleClick.bind(this);
    this.dispatchFileUpload = this.dispatchFileUpload.bind(this);
    this.handleUpload = this.handleUpload.bind(this);
    this.dispatchFileRemove = this.dispatchFileRemove.bind(this);
  }

  static getDerivedStateFromProps(nextProps) {
    // if the incoming profile image isn't in the default list then the chosen image will be the profile iamge
    // triggered on file upload
    if (nextProps.default_images.images.filter((e) => { return (e === nextProps.profile_image); }).length === 0)
    {
      //this.chosen_image = nextProps.profile_image;
      return { uploadLoader: "off", removeLoader: "off", message: "", chosen_image: nextProps.profile_image };
    } else {
      return null;
    }
  }

  handleClick(event) {
    // sets for the Dashboard state as a whole
    this.props.setProfileImage(event.target.value);
  }

  dispatchFileRemove() {
    this.props.setProfileImage("");
    this.setState({ removeLoader: "on" });
    document.getElementById("profileImageUpload").reset();
    //this.props.setActiveImage("");
  }

  dispatchFileUpload(e) {
    //console.log("DISPATCHING UPLOAD");
    document.getElementById("file").value = "";
    e.preventDefault();
    
    var event = new MouseEvent("click", {
      view: window,
      bubbles: true,
      cancelable: true
    });

    document.getElementById("file").dispatchEvent(event);
  }

  handleUpload(event) {
    //console.log("HANDLING UPLOAD:", event);
    this.props.freezeInput(true);
    this.setState({ uploadLoader: "on" });
    let msg = "";

    let self = this; // so this can get inside the then;
    let data = new FormData();
    data.append("image", event.target.files[0]);

    let postConfig = {
      method: "POST",
      url: "/dashboard/api/saveProfileImage",
      data: data
    };

    axios
      .request(postConfig)
      .then(function(res) {
        //console.log("SETTING PROFILE IMAGE ", res.data);
        self.props.setProfileImage(res.data.url);
        //UPDATE THUMBNAIL
      })
      .catch(error => {
        if (error.response) {
          // The request was made and the server responded with a status code
          // that falls out of the range of 2xx
          // console.log(error.response.data);
          // console.log(error.response.status);
          // console.log(error.response.headers);
          msg = "Only Accepts Image Files";
        } else if (error.request) {
          // The request was made but no response was received
          // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
          // http.ClientRequest in node.js
          msg = "";
        } else {
          // Something happened in setting up the request that triggered an Error
          msg = "Houston - We Have an Error!";
        }
        //console.log(error.config);
        this.setState({ uploadLoader: "off", message: msg });
        this.props.freezeInput(false);
      });
  }

  componentDidMount() {
    document.getElementById("file").addEventListener("change", this.handleUpload, false);
  }

  imageButtons(images, chosen_image) {
    let retval = [];
    images.map(img => {
      retval.push(<ProfileImageButton image={img} checked={chosen_image === img} handleClick={this.handleClick} key={img} />);
    });
    return retval;
  }

  render() {
    const images = this.props.default_images.images.concat([this.state.chosen_image]);

    return (
      <div className="row bottom-margin-lg">
        <div className="col-auto profileImage">
          {/*<span className="icon-upload-lg" />*/}
          <img
            className="icon-upload-lg"
            src={this.props.profile_image ? this.props.profile_image : "images/upload-icon.png"}
          />
        </div>
        <div className="col">
          {this.imageButtons(images, this.props.profile_image)}
          <div>
            <p>{this.props.recommendedSize}</p>

            <form id="profileImageUpload" action="/dashboard/api/saveProfileImage">
              <div className="row no-gutters">
                <button type="button" className="btn btn-primary proxyBtn" onClick={this.dispatchFileUpload}>
                  <div className="btn-loader">
                    <img className={"btn-loader-sm " + this.state.uploadLoader} src="images/loader.gif" />
                  </div>
                  {this.props.upload}
                </button>
                <button type="button" className="btn btn-primary" onClick={this.dispatchFileRemove}>
                  <div className="btn-loader">
                    <img className={"btn-loader-sm " + this.state.removeLoader} src="images/loader.gif" />
                  </div>
                  {this.props.remove}
                </button>
                <input type="file" name="image" className="hiddenBtn" id="file" />
              </div>
            </form>

            <div className="error-message">{this.state.message}</div>
          </div>
        </div>
      </div>
    );
  }
}

ProfileImage.propTypes = {
  profile_image: PropTypes.string,
  remove: PropTypes.string.isRequired,
  freezeInput: PropTypes.func,
  upload: PropTypes.string.isRequired,
  setProfileImage: PropTypes.func.isRequired,
  default_images: PropTypes.object.isRequired,
  recommendedSize: PropTypes.string
};

export default ProfileImage;
