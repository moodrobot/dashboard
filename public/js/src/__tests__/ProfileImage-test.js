import React, { Component } from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import ProfileImage from "../component/ProfileImage.js";
import ProfileImageButton from "../component/ProfileImageButton.js";
import translateData from "../../../../locales/en-US.json";

const file_upload = sinon.stub(ProfileImage.prototype, "componentDidMount");
const handle_click = sinon.spy(ProfileImage.prototype, "handleClick");

const default_images = {
  images: [
    "https://s3.amazonaws.com/dpi-prod/default-images/avatar-Autonomous-1.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/avatar-Autonomous-2.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/avatar-Data.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/avatar-GD.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/avatar-IoT-1.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/avatar-IoT-2.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/avatar-Realsense.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/bot01.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/one-eye.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/paper-bird.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/see-all-the-things.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/time.jpg",
    "https://s3.amazonaws.com/dpi-prod/default-images/wire-head.jpg"
  ],
  chosen_image: ""
};

describe("<ProfileImage />", () => {
  it("expect to match snapshot", () => {
    const wrapper = shallow(
      <ProfileImage
        profile_image="https://s3.amazonaws.com/dpi-intx/11531542/sirloin-1200-1200.jpg"
        setProfileImage={function() {}}
        default_images={default_images}
        recommendedSize={translateData["profile-image-recommended-size"]}
        remove={translateData["profile-image-remove"]}
        upload={translateData["profile-image-upload"]}
      />
    );

    /**EXPECT NO CHANGE IN THE SNAPSHOT */
    expect(wrapper).toMatchSnapshot();

    /** CHECK IMAGE/RADIO BUTTON CHANGE WHEN CLICKED */

    /*
    const image_button = wrapper.find(ProfileImageButton).first().dive();
    console.log(image_button.debug());
    image_button.find("input").first().simulate("click");//, {target: {value : default_images.default_img_01});
    wrapper.update();

    //CHECKS TO MAKE SURE HANDLE PROFILE IS CALLED WHEN CLICK NEW RADIO BTN IMAGE
    //expect(handle_click.called).toBe(true);
    //console.log(handle_click.called);
    //expect(handle_click.args[0][0].target.value).toBe(default_images.default_img_01);

    //console.log(wrapper.find("img").first().html());
    //expect(wrapper.find("input").first().props().value === profileData[0].first_name);

    /**ADD DIFFERENT PROPS TO SIMULATE CHANGE FROM ABOVE */
    //wrapper.setProps({data: profileData[1]});

    //wrapper.update();

    /** EXPECT PROPS TO UPDATE WITH NEW NAME */
    //expect(wrapper.find("input").first().props().value === profileData[1].first_name);
  });
});
