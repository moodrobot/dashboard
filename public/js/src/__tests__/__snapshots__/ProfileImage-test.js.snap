// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileImage /> expect to match snapshot 1`] = `
<div
  className="row bottom-margin-lg"
>
  <div
    className="col-auto profileImage"
  >
    <img
      className="icon-upload-lg"
      src="https://s3.amazonaws.com/dpi-intx/11531542/sirloin-1200-1200.jpg"
    />
  </div>
  <div
    className="col"
  >
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Autonomous-1.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Autonomous-1.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Autonomous-2.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Autonomous-2.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Data.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Data.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/avatar-GD.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/avatar-GD.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/avatar-IoT-1.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/avatar-IoT-1.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/avatar-IoT-2.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/avatar-IoT-2.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Realsense.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/avatar-Realsense.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/bot01.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/bot01.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/one-eye.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/one-eye.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/paper-bird.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/paper-bird.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/see-all-the-things.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/see-all-the-things.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/time.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/time.jpg"
    />
    <ProfileImageButton
      checked={false}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-prod/default-images/wire-head.jpg"
      key="https://s3.amazonaws.com/dpi-prod/default-images/wire-head.jpg"
    />
    <ProfileImageButton
      checked={true}
      handleClick={[Function]}
      image="https://s3.amazonaws.com/dpi-intx/11531542/sirloin-1200-1200.jpg"
      key="https://s3.amazonaws.com/dpi-intx/11531542/sirloin-1200-1200.jpg"
    />
    <div>
      <p>
        Recommended: 300x500 pixels
      </p>
      <form
        action="/dashboard/api/saveProfileImage"
        id="profileImageUpload"
      >
        <div
          className="row no-gutters"
        >
          <button
            className="btn btn-primary proxyBtn"
            onClick={[Function]}
            type="button"
          >
            <div
              className="btn-loader"
            >
              <img
                className="btn-loader-sm off"
                src="images/loader.gif"
              />
            </div>
            Upload
          </button>
          <button
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
          >
            <div
              className="btn-loader"
            >
              <img
                className="btn-loader-sm off"
                src="images/loader.gif"
              />
            </div>
            Remove
          </button>
          <input
            className="hiddenBtn"
            id="file"
            name="image"
            type="file"
          />
        </div>
      </form>
      <div
        className="error-message"
      />
    </div>
  </div>
</div>
`;
